README.md
<div align="center">

# GameGuard

GameGuard is a NodeJS game server that can be used to manage the players connecting to your game, manage rooms and the players in them, and more.

</div>

<div align="center">

  [![NPM version](https://img.shields.io/npm/v/gameguard.svg?style=flat)](https://www.npmjs.com/package/gameguard)
  [![Known Vulnerabilities](https://snyk.io/test/github/robertcorponoi/gameguard/badge.svg)](https://snyk.io/test/github/robertcorponoi/gameguard)
  ![npm](https://img.shields.io/npm/dt/gameguard)
  [![NPM downloads](https://img.shields.io/npm/dm/gameguard.svg?style=flat)](https://www.npmjs.com/package/gameguard)
  <a href="https://badge.fury.io/js/gameguard"><img src="https://img.shields.io/github/issues/robertcorponoi/gameguard.svg" alt="issues" height="18"></a>
  <a href="https://badge.fury.io/js/gameguard"><img src="https://img.shields.io/github/license/robertcorponoi/gameguard.svg" alt="license" height="18"></a>
  [![Gitter](https://badges.gitter.im/gitterHQ/gitter.svg)](https://gitter.im/robertcorponoi)

</div>

**Note:** This is the post 1.0.0 version of GameGuard that has lots of breaking changes from the last version due to major simplification. All of the previous features still exist but the API has changed to be more simple and streamlined. GameGuard can now be used on it's own but it has been simplified in order to be able to be extended further to suit your needs.

**Note About Logging:** For now, GameGuard has no logging capability. I've gone back and forth about implementing logging but I've found it to be so customizable it would be much easier for the end user to implement using signals but if there's enough requests then logging can be implemented to be a core part of GameGuard.

**Table of Contents**

- [Install](#install)
- [Initialization](#initialization)
- [Operation](#operation)
- [Docs](#docs)
  - [Database](#database)
  - [Players](#players)
  - [Rooms](#rooms)
  - [Global](#global)
- [Tests](#tests)

## **Install**

To install GameGuard you need the server side package (this one) and then a client-side package. Currently only [gameguard-client](https://github.com/robertcorponoi/gameguard-client) is supported but in the future there will be guides on creating your own client side solution to communicate with the GameGuard server.

To install GameGuard you can use:

```bash
$ npm install gameguard
```

and if you need gameguard-client, you can use:

```bash
$ npm install gameguard-client
```

**Note:** The documentation for gameguard-client will not be covered here but you can head over to the [gameguard-client documentation](https://github.com/robertcorponoi/gameguard-client#README.md) for client side usage.

## **Initialization**

To initialize GameGuard, you have to initialize it with a reference to a http or https server and an optional set of options.

| param                           | type        | description                                                                    | default                   |
|---------------------------------|-------------|--------------------------------------------------------------------------------|---------------------------|
| server                          | http.Server | A reference to the http server instance to bind to.                            |                           |
| options                         | Object      |                                                                                |                           |
| options.heartbeatInterval       | number      | The interval at which each player is pinged, in milliseconds.                  | 30000                     |
| options.latencyCheckInterval    | number      | The interval at which each player's latency is calculated, in milliseconds.    | 5000                      |
| options.maxLatency              | number      | The maximum latency, in milliseconds, the player can have before being kicked. | 300                       |
| options.mongodbConnectionString | string      | The connection string to use to connect to mongodb.                            | mongodb://localhost:27017 |

**Example:**

A basic example of initializing GameGuard this with my personal favorite http server, fastify, is as follows:

```js
'use strict'

const path = require('path');
const fastify = require('fastify')({ logger: true });

const GameGuard = require('gameguard');

const gg = new GameGuard(fastify.server);

fastify.listen(3000, (err, address) => {
  if (err) throw err;

  fastify.log.info(`server listening on ${address}`);
});
```

Here's an example of initializaing GamGuard with options:

```js
'use strict'

const path = require('path');
const fastify = require('fastify')({ logger: false });
const GameGuard = require('gameguard');

// Set the GameGuard server to use a latency check interval of 1000ms.
const gg = new GameGuard(fastify.server, { latencyCheckInterval: 1000 });

// Have the server listen on port 3000.
fastify.listen(3000, (err, address) => {
    if (err) throw err;
    console.log(`Listening on port 3000`);
});
```

Notice how we pass fastify's server instance to GameGuard so that GameGuard can use it to communicate with the client.

Let's also take a look how we can accomplish the same thing we did above but with express:

```js
'use strict'

const path = require('path');
const express = require('express');
const GameGuard = require('gameguard');

const app = express();

// Have the server listen on port 3000.
const server = app.listen(3000, () => console.log('Listening on port 3000'));

// Set the GameGuard server to use a latency check interval of 1000ms.
const gg = new GameGuard(server, { latencyCheckInterval: 1000 });
```

## **Operation**

Now let's talk about the operation of GameGuard:

1. The GameGuard server instance is created with a http server instance.

2. Now, the GameGuard server waits for a WebSocket connection from a page using the GameGuard client and in specific it waits for the client to send a message that contains the id of the player that connected.

3. Before the client becomes a player, the GameGuard server checks to see if the id of that client corresponds to a player in the database that is banned and if so their connection gets reject. Otherwise, the client is accepted and their player profile is created locally and updated in the database.

4. Everything is set up now. The player can be kicked, banned, messaged, or put into rooms.

## **Docs**

Since GameGuard is not a linear app and it's hard to go in order with what to document, we'll just go over the general aspect of each part of GameGuard and then link to the documentation for that module that goes in detail about it.

### **Database**

The GameGuard server uses the MongoDB to manage players with the mongoose package to manage the schemas and other operations. The database connection info has a default value of `mongodb://localhost:27017/gameguard` but you can configure the connection credentials in a `.env` file with a sample connection file provided as `.env.sample`. 

Check out the [database documentation](docs/database.md) for more specific database operations.

### **Players**

At it's core, GameGuard works around watching for clients trying to connect to the server and turning those clients into players. Once connected, players can be interacted with in forms of messaging, kicking, banning, or placing in rooms.

Check out the [player documentation](docs/player.md) for more specific player operations.

### **Rooms**

Rooms are used to group players together to perform similar actions together. For example, you can group players together in a room and easily send messages to all of them.

Check out the [room documentation](docs/room.md) for more specific room operations.

## **Global**

There are a few actions in GameGuard that are global and affect all players connected to the GameGuard server regardless of the rooms they are in.

Check out the [global documentation](docs/global.md) for more specific global operations.

## **Tests**

To run the tests for GameGuard server, you can use:

```bash
$ npm run test
```

## **License**

MIT


CHANGELOG.md
## 1.0.0 / 2020-09-26
- [FEATURE] Moved the player signals and methods out of `gameguard.players` and into `gameguard`.
- [FEATURE] Moved the room signals and methods out of `gmaeguard.rooms` and into `gameguard`.
- [FEATURE] Moved the system methods out of `gameguard.system` and into `gameguard`.
- [FEATURE] Removed support for mysql and streamlined the database operations.
- [FEATURE] Added convenience methods to add and update properties of players in the database.
- [TEST] Updated tests and removed unnecessary tests.
- [TEST] Updated framework tests and add npm scripts to run them.
- [DOCS] Updated documentation to reflect latest changes.
- [MISC] Updated out-of-date-dependencies to their latest versions.

## 0.14.1 / 2020-04-17
- [MISC] Updated out-of-date dependencies to their latest versions which also fixed all possible fixed security vulnerabilities.

## 0.14.0 / 2020-02-26
- [FEATURE] Made messages get send and read as binary.
- [MISC] Tested changes with latest version of gameguard-client.
- [MISC] Updated out-of-date dependencies to their latest versions. 

## 0.13.1 / 2020-02-16
- [HOTFIX] Made the initialization options have a default value if no options are provided.

## 0.13.0 / 2020-02-16
- [FEATURE] Added an option to auto-kick a player with a high latency according to a customizable `maxLatency` option.
- [FEATURE] Added a signal that gets dispatched when the player gets kicked for high latency.
- [DOCS] Added docs for new `maxLatency` option and the `timedOut` socket close message.
- [DOCS] Added docs for the timed out signal.
- [DOCS] Removed accidental typescript from code sample in README.
- [MISC] Updated dev dependencies to their latest versions.

## 0.12.0 / 2020-02-13
- [FEATURE] Added ability to customize close messages for the different events.
- [DOCS] Updated documentation for the feature added above.
- [MISC] Updated mongoose package to its latest version.

## 0.11.1 / 2020-02-11
- [DOCS] Updated docs to include information about the heartbeat and latency check.
- [MISC] Updated gameguard-client dev dependency to its latest version.

## 0.11.0 / 2020-02-11
- [FEATURE] Added a hearbeat that checks to see if clients are still connected to the server or not on a customizable interval.
- [FEATURE] Added a latency check to get the players latency on a customizable interval.
- [DOCS] Fixed random old documentation.
- [MISC] Updated dependencies to their latest versions.
- [MISC] Changed CHANGELOG formatting.
- [MISC] Removed unnecessary dependencies.

## 0.10.0 / 2020-01-28
- [FEATURE] Normalized signal names.
- [FEATURE] Changed rooms `created` to `rooms` and players `connected` to `players` since signal names changed.
- [DOCS] Updated documentation for new signal names.
- [DOCS] Added individual player documentation at docs/player.md.
- [TEST] Udpated tests.

## 0.9.0 / 2020-01-28
- [FEATURE] Changed from using events to signals.
- [DOCS] Updated documentation for signals.
- [TEST] Updated tests.
- [MISC] Updated dependencies to their latest versions.

## 0.8.0 / 2020-01-25
- [FEATURE] Added option to use mysql.
- [FEATURE] Changed the `database` option to `dbType` for more clarity.
- [FEATURE] Added option to use a .env file to specify database connection information.
- [DOCS] Added a .sample.env file to highlight the options available for use in a .env file and their default values.
- [DOCS] Updated README with lateast information.

## 0.7.0 / 2020-01-19
- [FEATURE] Matched version with gameguard-client. From this point forward both packages with matching version numbers will be guaranteed to work with each other.

## 0.5.1 / 2020-01-18
- [FEATURE] Added mongodb as the default storage method.
- [BREAKING] Local storage is broken as of this patch and so mongodb is the only option until the next storage update when support for mysql is added.
- [MISC] Added breaking change note to README.

## 0.5.0 / 2020-01-08
- [FEATURE] Moved the client-side component to its own package, see the README for more.
- [FEATURE] Messages now require a message object instead of type and content.
- [FEATURE] Changed the `player-joined` event to `player-connected`.
- [FEATURE] Changed the `player-left` event to `player-disconnected`.
- [FEATURE] Converted the project to typescript.
- [FEATURE] Changed the `players` property of the players module to be `connected`.
- [FEATURE] Changed the `rooms` property of the rooms module to `created`.
- [MISC] Removed unnecessary dependencies and moved express to a be a dev dependency which it should have been all along.
- [MISC] Updated README to reflect breaking changes.

## 0.4.3 / 2020-01-08
- [MISC] Made private methods that weren't really private public.
- [MISC] Updated dependencies to their latest versions.

## 0.4.2 / 2020-01-05
- [MISC] Slightly cleaned up docs in preparation for new functionality.

## 0.4.1 / 2020-01-05
- [HOTFIX] Added main property to package.json.
- [MISC] Updated license year to reflect new year.
- [MISC] Removed extra spacing.
- [MISC] Removed debugging statements.
- [MISC] Updated dependencies to their latest versions.

## 0.4.0 / 2019-12-24
- [MISC] Updated tests to clear database file before each one.
- [MISC] Added more test cases.
- [MISC] Moved all tests to tests folder.
- [HOTFIX] Changed the client to use window.location.host instead of localhost.
- [HOTFIX] Made secure flag change connection string to wss.

## 0.3.0 / 2019-12-24
- [FEATURE] Changed class properties to be strict and added the necessary setters/getters to be more strict.
- [MISC] Updated JSdoc comments.
- [MISC] Added badges to README and updated getters/setters documentation.
- [MISC] Updated dependencies to their latest versions.

## 0.1.0 / 2019-10-16
- Initial release


gameguardclient.js
/**
 * Represents a message sent by the client to the server.
 * 
 * This provides structure for messages sent along with metadata that can be helpful to the server.
 */
class Message {

  /**
   * @param {string} type The type of message that is being sent.
   * @param {string} content The actual contents of the message.
   */
  constructor(type, content) {

    /**
     * The type of message being sent.
     * 
     * @property {string}
     */
    this.type = type;

    /**
     * The actual contents of the message.
     * 
     * @property {string}
     */
    this.content = content;

    /**
     * The timestamp of when this message was created and sent.
     * 
     * @property {number}
     */
    this.timestamp = + new Date();

  }

  /**
   * Prepare this message to be sent by stringifying the contents of it.
   * 
   * @since 0.1.0
   * 
   * @returns {string} Returns the stringified version of this message.
   */
  stringify() {

    const message = { type: this.type, content: this.content, timestamp: this.timestamp };

    return JSON.stringify(message);

  }

}

/**
 * The options available and their default values for an instance of the GameGuard client.
 */
class Options {

  /**
   * @param {Object} [options={}] The initialization parameters passed to the GameGuard client instance. 
   * @param {boolean} [options.secure=false] Indicates whether the websocket will connect to the server with a secure connection or not.
   */
  constructor(options) {

    /**
     *  Indicates whether the websocket will connect to the server with a secure connection or not.
     * 
     * @property {boolean}
     * 
     * @default false
     */
    this.secure = false;

    /**
     * Override the default options with user specified ones if they exist.
     */
    Object.assign(this, options);

  }

}

/**
 * Generates various types of uuids to specifications.
 */
var uuid = {

  /**
   * Generates a v4 compliant uuid.
   * 
   * This is derived from this post: https://stackoverflow.com/a/2117523/4274475
   * 
   * @returns {string} Returns a valid v4 uuid.
   */
  v4() {

    return ([1e7] + -1e3 + -4e3 + -8e3 + -1e11).replace(/[018]/g, c =>
      (c ^ crypto.getRandomValues(new Uint8Array(1))[0] & 15 >> c / 4).toString(16)
    );

  }

};

/**
 * Provides methods to get, set, or edit data in cookies.
 */
var cookies = {

  /**
   * Gets the value for a specified cookie name.
   * 
   * @param {string} name The name of the cookie to get.
   * 
   * @returns {string} Returns the value of the cookie or an empty string if the cookie does not exist.
   */
  get(name) {

    const cname = `${name}=`;

    const decodedCookie = decodeURIComponent(document.cookie);

    const ca = decodedCookie.split(';');

    for (let c of ca) {

      while (c.charAt(0) == ' ') c = c.substring(1);

      if (c.indexOf(cname) == 0) return c.substring(cname.length, c.length);

    }

    return '';

  },

  /**
   * Sets a new cookie with the desired name, value, and expiration date in days.
   * 
   * @param {string} name The name of the cookie to set.
   * @param {string} value The value of the cookie to set.
   * @param {number} daysToExpire The number of days until this cookie expires.
   */
  set(name, value, daysToExpire) {

    const d = new Date();

    d.setTime(d.getTime() + (daysToExpire * 24 * 60 * 60 * 1000));

    const expires = `expires=${d.toUTCString()}`;

    document.cookie = `${name}=${value};${expires};path=/`;

  }

};

/**
 * Provides methods to set, get, and work with data stored on the client side in cookies.
 */
class ClientData {

  /**
   * Gets the GameGuard player id from the current player, if it doesn't exist then a player id will
   * be assigned to the player and returned.
   * 
   * @returns {string} Returns the player id.
   */
  getPlayerId() {

    let playerId = cookies.get('gameguardPlayerId');

    if (!playerId) {

      playerId = uuid.v4();

      cookies.set('gameguardPlayerId', playerId);

    }

    return playerId;

  }

}

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

var Listener =
/**
 * The function that will be called when the listener is processed.
 * 
 * @property {Function}
 */

/**
 * The context to use when calling this listener.
 * 
 * @property {*}
 */

/**
 * Whether or not this listener will be automatically destroyed after being run once.
 * 
 * @property {boolean}
 */

/**
 * Keeps track of the number of times that this listener has been called.
 * 
 * @property {number} 
 */
function Listener(fn, ctx, once) {
  _classCallCheck(this, Listener);

  _defineProperty(this, "fn", void 0);

  _defineProperty(this, "ctx", void 0);

  _defineProperty(this, "once", void 0);

  _defineProperty(this, "timesCalled", 0);

  this.fn = fn;
  this.ctx = ctx;
  this.once = once;
};

/**
 * Compare two functions by turning them into strings and removing whitespace/line-breaks and then checking equality.
 * 
 * @param {Function} fn1 The first function.
 * @param {Function} fn2 The second function.
 * 
 * @returns {boolean} Returns true if the functions are equal and false otherwise.
 */

function compareFunctions(fn1, fn2) {
  var f1 = fn1.toString().replace(/\n/g, '').replace(/\s{2}/g, ' ');
  var f2 = fn2.toString().replace(/\n/g, '').replace(/\s{2}/g, ' ');
  if (f1 === f2) return true;
  return false;
}

/**
 * Eventverse is a higly performant and easy to use event emitter for Nodejs and the browser.
 */

var Eventverse =
/*#__PURE__*/
function () {
  /**
   * The maximum amount of listeners each event can have at one time.
    * 
    * @private
   * 
   * @property {number}
   * 
   * @default 10
   */

  /**
   * A collection of all of the listeners created for this instance of Eventverse.
   * 
   * @property {Object}
   */

  /**
   * @param {number} [maxListenerCount=10] The maximum amount of listeners each event can have at one time. 
   */
  function Eventverse() {
    var maxListenerCount = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 10;

    _classCallCheck(this, Eventverse);

    _defineProperty(this, "_maxListenerCount", void 0);

    _defineProperty(this, "events", Object.create(null));

    this._maxListenerCount = maxListenerCount;
  }
  /**
   * Returns the number of max listeners each event can have at one time.
   * 
   * @returns {number}
   */


  _createClass(Eventverse, [{
    key: "listenerCount",

    /**
     * Returns the number of listeners for a given event.
     * 
     * @param {string} event The name of the event.
     * 
     * @returns {number}
     */
    value: function listenerCount(event) {
      return this.events[event].length;
    }
    /**
     * Returns the number of times a listener was called.
     * 
     * @param {string} event The name of the event to get the times called for.
     * 
     * @returns {number} Returns the number of times the event was called.
     */

  }, {
    key: "timesCalled",
    value: function timesCalled(event) {
      return this.events[event][0].timesCalled;
    }
    /**
     * Runs all of the listeners attached to this Eventverse with the event name and with the supplied arguments.
     * 
     * @param {string} event The name of the event to emit.
     * @param {...*} args The arguments to pass to the listeners.
     */

  }, {
    key: "emit",
    value: function emit(event) {
      if (!this._exists(event)) return;
      var listeners = this.events[event];

      for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        args[_key - 1] = arguments[_key];
      }

      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = listeners[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var _listener$fn;

          var listener = _step.value;

          (_listener$fn = listener.fn).call.apply(_listener$fn, [listener.ctx].concat(args));

          listener.timesCalled++;
          if (listener.once) this.removeListener(event, listener.fn);
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator["return"] != null) {
            _iterator["return"]();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }
    }
    /**
     * Adds a listener function for the given event.
     * 
     * 
     * @param {string} event The name of the event to add a listener for.
     * @param {Function} fn The function to run when the event is emitted.
     * @param {Object} context The context to use when calling the listener.
     * @param {boolean} once Indicates whether this listener should only be called once.
     * 
     * @returns {Eventverse} Returns this for chaining.
     */

  }, {
    key: "addListener",
    value: function addListener(event, fn) {
      var context = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : this;
      var once = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : false;
      var listener = new Listener(fn, context, once);

      if (!this._exists(event)) {
        this.events[event] = [];
      } else if (this.events[event].length === this.maxListenerCount) {
        console.warn("[Eventverse][addListener]: The event ".concat(event, " already has the max amount of listeners."));
        return;
      }

      this.events[event].push(listener);
      return this;
    }
    /**
     * Removes a listener function for the given event.
     * 
     * @param {string} event The name of the event to remove the listener on.
     * @param {Function} listener The listener to remove from the event.
     * 
     * @returns {Eventverse} Returns this for chaining.
     */

  }, {
    key: "removeListener",
    value: function removeListener(event, listener) {
      var _this = this;

      if (!this._exists(event)) {
        console.warn('[Eventverse][removeListener]: Unable to remove listener for an event that doesnt exist.');
        return;
      }

      var _iteratorNormalCompletion2 = true;
      var _didIteratorError2 = false;
      var _iteratorError2 = undefined;

      try {
        var _loop = function _loop() {
          var eventListener = _step2.value;

          if (compareFunctions(eventListener.fn, listener)) {
            _this.events[event] = _this.events[event].filter(function (evListener) {
              return evListener != eventListener;
            });
            return "break";
          }
        };

        for (var _iterator2 = this.events[event][Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
          var _ret = _loop();

          if (_ret === "break") break;
        }
      } catch (err) {
        _didIteratorError2 = true;
        _iteratorError2 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion2 && _iterator2["return"] != null) {
            _iterator2["return"]();
          }
        } finally {
          if (_didIteratorError2) {
            throw _iteratorError2;
          }
        }
      }

      return this;
    }
    /**
     * Removes all listeners from a given event.
     * 
     * @param {string} event The name of the event to remove all listeners from.
     * 
     * @returns {Eventverse} Returns this for chaining.
     */

  }, {
    key: "removeAllListeners",
    value: function removeAllListeners(event) {
      if (!this._exists(event)) {
        console.warn('[Eventverse][removeAllListeners]: Unable to remove listener for an event that doesnt exist.');
        return;
      }

      this.events[event] = [];
      return this;
    }
    /**
     * Add a listener function that will only run once.
     * 
     * @param {string} event The name of the event to add a listener for.
     * @param {Function} fn The function to run when the event is emitted.
     * @param {Object} [context=this] The context to use when calling the listener.
     * 
     * @returns {Eventverse} Returns this for chaining.
     */

  }, {
    key: "once",
    value: function once(event, fn) {
      var context = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : this;
      this.addListener(event, fn, context, true);
      return this;
    }
    /**
     * Adds a listener function for the given event.
     * 
     * @param {string} event The name of the event to add a listener for.
     * @param {Function} fn The function to run when the event is emitted.
     * @param {Object} [context=this] The context to use when calling the listener.
     * 
     * @returns {Eventverse} Returns this for chaining.
     */

  }, {
    key: "on",
    value: function on(event, fn) {
      var context = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : this;
      this.addListener(event, fn, context);
      return this;
    }
    /**
     * Checks if an event exists.
      * 
     * @private
     * 
     * @param {string} event The name of the event.
     * 
     * @returns {boolean} Returns true if the event exists or false otherwise.
     */

  }, {
    key: "_exists",
    value: function _exists(event) {
      if (this.events[event]) return true;
      return false;
    }
  }, {
    key: "maxListenerCount",
    get: function get() {
      return this._maxListenerCount;
    }
  }]);

  return Eventverse;
}();

/**
 * The GameGuard client is used to establish a connection to the server send player info.
 */
class GameGuardClient extends Eventverse {

  /**
   * @param {Object} [options] The initialization parameters passed to this instance.
   * @param {boolean} [options.secure=false] Indicates whether the websocket will connect to the server with a secure connection or not.
   */
  constructor(options = {}) {

    super();

    /**
     * A reference to this client's options.
     * 
     * @private
     * 
     * @property {Options}
     */
    this._options = new Options(options);

    /**
     * A reference to this client's WebSocket connection.
     * 
     * @private
     * 
     * @property {WebSocket}
     */
    this._socket = null;

    /**
     * A reference to the ClientData module.
     * 
     * @private
     * 
     * @property {ClientData}
     */
    this._clientData = new ClientData();

    this._boot();

  }

  /**
   * Initialize the WebSocket connection and all of the events that we need to respond to.
   * 
   * @private
   */
  _boot() {

    const wsProtocol = this._options.secure ? 'wss' : 'ws';

    this._socket = new WebSocket(`${wsProtocol}://${window.location.host}/`);

    this._socket.addEventListener('open', () => this._onOpen());

    this._socket.addEventListener('message', (message) => this._onMessage(message));

    this._socket.addEventListener('close', (ev) => this._onClose(ev));

  }

  /**
   * When the WebSocket connection opens, we check to see if they are an existing GameGuard player using cookies
   * and if they are not then they are assigned a new GameGuard player id.
   * 
   * @private
   */
  _onOpen() {

    const playerId = this._clientData.getPlayerId();

    const message = new Message('player-joined', playerId);

    this._socket.send(message.stringify());

    this.emit('open', playerId);

  }

  /**
   * TODO:
   * 
   * @param {string} message The message Object received from the server.
   */
  _onMessage(message) {

    console.log('blah');

    const parsed = JSON.parse(message.data);

    const msg = new Message(parsed.type, parsed.content);

    this.emit('message', msg);

  }

  /**
   * When the WebSocket connection closes, we end the players connection to the game and notify them why, if a reason
   * was provided.
   * 
   * @property {Event} ev The WebSocket close event Object.
   */
  _onClose(ev) {

    this.emit('close', ev);

  }

}

export default GameGuardClient;
